### ISSUE TITLE
`@codec.decode` should support defining metadata.

### ISSUE THREAD

**Issue by ph** [elastic/logstash#4289]
I have encountered an issue with the lumberjack and the beats plugin combined with the multiline codec and its really hard to fix without changing the API. The problem is related to spooling or the buffering of events within the codec code. 

Some protocol support adding useful metadata (lumberjack, rabbitmq) to an event, but the `decode` API only work with string. It look like an async api but its more like a sync api.

See the following example:

```ruby
raw_string = "bonjour la famille"
data = {  "host" => "sashimi", "line" => raw_string }
@codec.decode(data) do |event|
   event["host"] = data["host"]
end
```

This code is simple and its always right if the codec doesn't do buffering, since the proc will capture the data hash.

If the codec do some kind of buffering like the multiline filter it will only be right if new lines trigger a flush, in any other context like time based eviction or when we manually trigger a flush before shutting down the plugin that context will be lost. 

We need revisit that api to add the possibility to add context/or metadata to the decode, or provide an object that can receive callbacks. Until then I have implemented the eviction block to only decorate and push to the queue the events, but the metadata wont be added to them :(

This issue is related to @guyboertje proposal for changing the interaction between the codec and the queues.

**Comment by ph** [comment]
I would like to be able to send an object to the decode method that the codec could use to do some signaling or do more manipulation.

```ruby
class CustomCodecAction
  def initialize(metadata)
    @metadata = metadata
  end

  def on_create(event)
    event.merge(@metadata)
  end
end

CustomCodecAction.new(data)

@codec.decode(string, CustomCodecAction.new(data)) do .. end

**Comment by ph** [comment]
Maybe we could do theses changes in a backward compatible way with the current codec and only implement a proxy.

**Comment by wrigby** [comment]
@ph I was about to open an issue saying exactly the same thing, but noticed you had already opened this one. In the case of the Netflow codec, it's actually very important for the codec to know the source IP address of the router sending netflow reports. Unfortunately, there's no way for the input plugin to pass that information along to the codec (see logstash-plugins/logstash-codec-netflow#9).

:+1: for the ability to pass metadata from an input to a codec.

**Comment by jorritfolmer** [comment]
Yes seconded! Access to metadata (src_ip and src_port in this case) would make the logstash-codec-netlfow work again in all the use-cases in which it now fails. Suppose you export Netflow from a Cisco ASA firewall and a Cisco router (logstash-plugins/logstash-codec-netflow#21): can't be done.
Suppose you use different softflowd versions on Linux or BSD: can't be done either.

That's because the information to decode the netflow data is only sent once in a while, and there isn't enough data in the netflow payload itself to keep track of which decoding packet belongs to which data packet. So for example the decoding information for softflowd from box A is applied to netflow data from box B, leading to failure or garbage data. The codec needs to keep track from which exporter the decoding information came from to properly decode.

**Comment by ph** [comment]
@guyboertje is working on improving how we work with the codec from the inputs, this is all good information.

**Comment by jorritfolmer** [comment]
Any update on this?  (Also PR logstash-plugins/logstash-input-udp#20, but please don't laught)

**Comment by guyboertje** [comment]
@jorritfolmer - work on this fundamental change has been suspended while I work on queue persistence.

One can think of an input as a orchestration mechanism with a data source loop stage and an event generation stage.

That said, my design centred around the idea of getting a context hash + the raw data from the data source, allowing the data source to put the raw data in context before it gives it to the event generation stage.

We are planning to move the what is currently a codec to being a filter i.e. after the queue - because it is a design goal to minimise the amount of processing done to raw data before it is persisted to disk. We are doing this for two reasons: 1) to move the processing overhead from the limited threads available in an input to the more scalable worker threads, 2) to lessen the risk of data loss in the inputs due to bugs or data incompatibility. **However**, only the codecs that are stateless can be moved. Ones that hold state or buffer raw data cannot be moved and we must find a way to accommodate these in the event generation stage.

If I understand you correctly to fix issue 21 in the netflow codec it will need to be stateful - it will need to cache the template in when seen against some identity (ip and port) so that it can apply that template to  data it sees subsequently, correct.

If this is true, you are better off creating a netflow input that includes the current codec logic directly in the input - you then have access to any context directly. It can be based on the udp input.

This new input will be more compatible with the next generation of LS as well.

I can help you if you decide to do this.

**Comment by jorritfolmer** [comment]
Thanks for taking the time to explain that a new `logstash-input-netflow` is the way to go, since it does indeed need to keep state. I'll look into it the coming weeks.

**Comment by guyboertje** [comment]
@jorritfolmer - you are most welcome.
